File: src/main/resources/js/pyramid.js
function pyramidIterative(num) {

    var strArr = [];
    var stringSize = (2 * num) - 1;
    var midIdx = Math.floor(stringSize / 2);

    for (var i = 0; i < num; i++) {
        var chars = [];
        // In ES6 we'd use '#'.repeat(i + 1), but Nashorn is only 5.1
        for (var j = 0; j < stringSize; j++) {
            chars.push(j < midIdx - i ? ' ' : (j >= midIdx + 1 + i ? ' ' : '#'));
        }
        strArr.push(chars.join(''));
    }

    return Java.to(strArr, 'java.lang.String[]');
}

function pyramidRecursive(num) {

    return Java.to(pyramidHelper([], num), 'java.lang.String[]');
}

function pyramidHelper(strArr, num) {

    if (num <= 0) {
        return strArr;
    }

    var stringSize = (2 * (num + strArr.length)) - 1;
    var midIdx = Math.floor(stringSize / 2);

    var chars = [];
    for (var j = 0; j < stringSize; j++) {
        chars.push(j < midIdx - strArr.length ? ' ' : (j >= midIdx + 1 + strArr.length ? ' ' : '#'));
    }
    strArr.push(chars.join(''));

    return pyramidHelper(strArr, num - 1);
}

File: src/main/java/com/devxpress/Pyramid.java
package com.devxpress;

import java.util.Arrays;

public class Pyramid {

    public static String[] pyramidIterative(int num) {

        String[] results = new String[num];
        int stringSize = (2 * num) - 1;
        int midIdx = stringSize / 2;

        for (int i = 0; i < num ; i++) {
            char[] chars = new char[stringSize];
            Arrays.fill(chars, 0, midIdx - i, ' ');
            Arrays.fill(chars, midIdx - i, midIdx + 1 + i, '#');
            Arrays.fill(chars, midIdx + 1 + i, stringSize, ' ');
            results[i] = String.valueOf(chars);
        }

        return results;
    }

    public static String[] pyramidRecursive(int num) {

        return recursionHelper(new String[num], num);
    }

    private static String[] recursionHelper(String[] strArr, int num) {
        if (num <= 0) {
            return strArr;
        }

        int arrSize = getLength(strArr);
        int stringSize = (2 * (num + arrSize)) - 1;
        int midIdx = stringSize / 2;

        char[] chars = new char[stringSize];
        Arrays.fill(chars, 0, midIdx - arrSize, ' ');
        Arrays.fill(chars, midIdx - arrSize, midIdx + 1 + arrSize, '#');
        Arrays.fill(chars, midIdx + 1 + arrSize, stringSize, ' ');
        strArr[arrSize] = String.valueOf(chars);

        return recursionHelper(strArr, num - 1);
    }

    private static int getLength(String[] arr) {
        int count = 0;
        for (String str : arr) {
            if (str != null) {
                count++;
            }
        }

        return count;
    }

}
